core/store/src/adapter/trie_store.rs

```diff
@@ -98,23 +99,23 @@ impl<'a> TrieStoreUpdateAdapter<'a> {
         Self { store_update: StoreUpdateHolder::Reference(store_update) }
     }
 
-    fn get_key_from_shard_uid_and_hash(&self, shard_uid: ShardUId, hash: &CryptoHash) -> [u8; 40] {
-        get_key_from_shard_uid_and_hash(&self.store_update.store, shard_uid, hash)
-    }
-
     pub fn decrement_refcount_by(
         &mut self,
         shard_uid: ShardUId,
         hash: &CryptoHash,
         decrement: NonZero<u32>,
     ) {
-        let key = self.get_key_from_shard_uid_and_hash(shard_uid, hash);
-        self.store_update.decrement_refcount_by(DBCol::State, key.as_ref(), decrement);
+        // For Resharding V3, along with decrementing the refcount of the child shard_uid, we also need to
+        // decrement the refcount of the parent shard_uid, if it exists.
+        let mapped_shard_uid = maybe_get_shard_uid_mapping(&self.store_update.store, shard_uid);
```

Comment:
Sanity Check - The mapped shard uid may be the parent or an older ancestor. Does this solution for for the case with ancestors other than parent?

Reply:
I think the new assumption is that we never map to an older ancestor, because the mapping is removed within an epoch after resharding.

Reply:
That's not entirely true for archival storage, as we would still have to maintain ancestor mapping. This would add some overhead with refcount increments but archival nodes don't have refcount decrements.

Reply:
Could you expand on that? Don’t we also remove the mapping on archival nodes once the child state is created? If so, doesn’t that mean we would never map to the grandparent in the new version?

Reply:
Apologies, you're right, we would remove the mapping for archival nodes as well :)